accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+   directly to the underlying Wasm being executed. This provides Wasm-level
+   debugging to the developer.
+
+2. **The DWARF Debug Adapter:** A debug adapter component that translates
+   queries from the debugger through the source program’s DWARF debug info
+   custom sections. This enables the debugging of programs compiled to Wasm,
+   such as Rust, C, C++, etc… 
+
+Interpreters that are compiled to Wasm can, in the future, provide their own
+debug adapter components that translate debugging queries through their runtime
+data structures to inspect debuggee programs being interpreted within
+themselves.
+
+Even further into the future, we can collaborate with the WebAssembly Community
+Group to standardize debug adapter components (or something evolved from them)
+in the ``WebAssembly/tool-conventions`` and `WebAssembly/component-model`
+repositories.
+
+While the exact WIT interface of and world for debug adapter components is out
+of scope for this RFC, and will require experimentation and prototyping during
+implementation, we expect that it will closely mirror the kinds of queries
+supported by the Debug Adapter Protocol.
+
+## Live Debugging
+
+Live debugging support comprises several features, each of which comes with
+design and implementation considerations. These features are: setting
+breakpoints, single stepping, watch points, step into/out of, and integration
+with our existing baseline compiler Winch.
+
+### Breakpoints and Single-stepping
+
+To implement breakpoints and single-stepping, we need the ability to pause
+program execution and wait for user input prior to executing a specific line of
+code. The way this is handled in native debuggers is as follows:
+
+* When a user sets a breakpoint, the instruction corresponding to that line of
+  code is replaced with an `int3` instruction.
+
+* When executed, the `int3` instruction raises a `SIGTRAP`, halting execution
+  and yielding control back to the debugger.
+
+This approach leads to a problem for us: we only have one compiled copy of a
+module’s code, even though multiple instances may be running the code. So if we
+just replace an instruction in the code with an interrupt, it is possible that a
+different instance will hit that interrupt. At that point, in order for that
+instance to continue executing we need to swap in the actual instruction. This
+leads to a race condition between the instances where we aren’t guaranteed that
+our breakpoint will actually interrupt when we need it to! We can get around
+this in the following way:
+
+1. When code is compiled for debugging, for every instruction emitted, emit a
+call to a utility function that can check if execution should be paused:
+
+2. When a breakpoint is set, the function will pause execution, allowing for
+resumption later on
+
+3. When an instance hits that breakpoint it can automatically continue if it
+isn’t intended to stop, otherwise it can wait for user input.
```

Comment:
Another option is to still use int3, but when the breakpoint is hit, interrupt all other instances that use the same module, swap the int3 away, single step the instance which shouldn't have hit it and then insert the int3 again and resume all other instances. This only slows down execution when a breakpoint is actually hit.

Reply:
I agree with @bjorn3 here -- this can be eventually implemented as an optimization in a second pass, as I'm sure we at first mostly want to see this whole contraption working.

I'd like to see, though, in the Future Work section, something that allows implementing platform-specific debugging facilities (e.g. using `ptrace(2)` on Unix-like systems), and leaving this approach as a baseline implementation for those host platforms where we don't yet implement them.

Reply:
Some host environments will have constraints about modifying executable code, so starting with a strategy that doesn't involve post-compilation code mutation would be ideal for maximizing the portability of debugging.

Another approach I've seen used (e.g., in Firefox's baseline JIT when debugging is enabled) is to, before every instruction, compile an inline branch on a memory location (set when stepping is enabled) that branches around the call to the debug handler function.  This ends up being fairly inexpensive when stepping is not active.  When stepping is active, you can then optimize the overhead as much as desired by having a separate testable memory location per { instance, function, instruction }.